trillian/util/log_leaf.go

```diff
@@ -71,3 +50,57 @@ func ExtraDataForChain(cert ct.ASN1Cert, chain []ct.ASN1Cert, isPrecert bool) ([
 	}
 	return tls.Marshal(extra)
 }
+
+func BuildLogLeafWithHash(logPrefix string, merkleLeaf ct.MerkleTreeLeaf, leafIndex int64, cert ct.ASN1Cert, chain []ct.ASN1Cert, chainHash []byte, isPrecert bool) (trillian.LogLeaf, error) {
+	return buildLogLeaf(logPrefix, merkleLeaf, leafIndex, cert, chain, chainHash, isPrecert)
+}
+
+// ExtraDataForChainWithHash creates the extra data associated with a log entry as
+// described in RFC6962 section 4.6.
+func ExtraDataForChainWithHash(cert ct.ASN1Cert, chain []ct.ASN1Cert, chainHash []byte, isPrecert bool) ([]byte, error) {
+	var extra interface{}
+
+	if isPrecert {
+		// For a pre-cert, the extra data is a TLS-encoded PrecertChainEntry.
+		extra = ct.PrecertChainEntryHash{
+			PreCertificate:    cert,
+			IssuanceChainHash: chainHash,
+		}
+	} else {
+		// For a certificate, the extra data is a TLS-encoded:
+		//   ASN.1Cert certificate_chain<0..2^24-1>;
+		// containing the chain after the leaf.
+		extra = ct.CertificateChainHash{
+			IssuanceChainHash: chainHash,
+		}
+	}
+	return tls.Marshal(extra)
+}
+
+func buildLogLeaf(logPrefix string, merkleLeaf ct.MerkleTreeLeaf, leafIndex int64, cert ct.ASN1Cert, chain []ct.ASN1Cert, chainHash []byte, isPrecert bool) (trillian.LogLeaf, error) {
```

Comment:
Can you leave a comment explaining that chainHash controls some logic? Specifically, if `chainHash` is not nil, but neither is `chain`, then `chain` will be ignored.

As an aside, is it possible for `chainHash` to be nil just because the chain is empty, even if `extra_data_issuance_chain_storage_backend` was set to mysql?

Reply:
Done.

If the chain is empty, there will still be a `chainHash`. See the following example.
https://go.dev/play/p/_ZCJHlvc5ZF